### PR TITLE
{bio}[goolf-1.7.20] Ghostscript 9.16 (REVIEW)

### PR DESCRIPTION
**Pull request by npklein**
depends on ~~#1704~~, ~~#1710~~, ~~#1711~~, ~~#1712~~, ~~#1713~~, ~~#1714~~, ~~#1715~~, ~~#1716~~

### CODE REVIEW

**Comment by boegel** [comment]
Jenkins: ok to test

**Comment by boegel** [comment]
@npklein: please include a line stating that this PR depends on another (like I did now)

File: easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ different targets. It used to be part of the cups printing stack, but is no longer used for that."""
+
+toolchain = {'name': 'goolf', 'version': '1.7.20'}
+toolchainopts = {'pic': True}
+
+source_urls = ["http://downloads.ghostscript.com/public/"]
+sources = [SOURCELOWER_TAR_GZ]
+
+# expat, freetype and libpng are optional deps,
+# these are actually patched and included in the sources,
```
Comment (boegel) [inline easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb]: a recent easyconfig does include the deps here, cfr. https://github.com/hpcugent/easybuild-easyconfigs/blob/master/easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.14-intel-2014b.eb

So it seems like that's a viable option?

**Comment by npklein** [comment]
depends on freetype/2.5.3-goolf-1.7.20, fontconfig/2.11.1-goolf-1.7.20, GLib/2.40.0-goolf-1.7.20, libjpeg-turbo/1.3.1-goolf-1.7.20, libpng/1.6.12-goolf-1.7.20, expat/2.1.0-goolf-1.7.20, and cairo/1.12.18-goolf-1.7.20

**Comment by boegel** [comment]
@npklein: please include references to the PRs providing these (in the description)

File: easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+source_urls = ["http://downloads.ghostscript.com/public/"]
+sources = [SOURCELOWER_TAR_GZ]
+
+dependencies = [
+    ('freetype', '2.5.3'),
+    ('fontconfig', '2.11.1'),
+    ('GLib', '2.40.0'),
+    ('libjpeg-turbo', '1.3.1'),
+    ('libpng', '1.6.12'),
```
Comment (fgeorgatos) [inline easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb]: The above would be nice to become aligned with common_deps #1689.
ie. bump forward libpng; or, does that bring up dependency hell with other deps? if so, it's a pass from me...

Comment (boegel) [inline easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb]: +1

**Comment by npklein** [comment]
@boegel I get an AssertionError: No conflicts detected, but when I sync this branch with develop I don't get a merge error. Do you know what is conflicting?

**Comment by boegel** [comment]
@npklein: this is not a git conflict, but a conflict on the versions of `libpng` required in the `Ghostscript` dependency tree, in this case...

one of the `Ghostscript` dependencies required libpng 1.6.12, while Ghostscript itself asks for libpng 1.6.17...

**Comment by npklein** [comment]
@fgeorgatos thanks, it were only a few dependencies to check so was easily found

File: easybuild/easyconfigs/f/freetype/freetype-2.5.3-goolf-1.7.20.eb [hunk 3]
```diff
@@ -11,7 +11,7 @@ toolchain = {'name': 'goolf', 'version': '1.7.20'}
 source_urls = [GNU_SAVANNAH_SOURCE]
 sources = [SOURCE_TAR_GZ]
 
-dependencies = [('libpng', '1.6.12')]
+dependencies = [('libpng', '1.6.17')]
```
Comment (boegel) [inline easybuild/easyconfigs/f/freetype/freetype-2.5.3-goolf-1.7.20.eb]: we try to stay away from bumping dependency versions in existing easyconfigs...

File: easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+source_urls = ["http://downloads.ghostscript.com/public/"]
+sources = [SOURCELOWER_TAR_GZ]
+
+checksums = ['818c87e31f7562aaa97397d3d0cc20a1']
+
+dependencies = [
+    ('zlib', '1.2.8'),
+    ('libpng', '1.6.17'),  
+    ('freetype', '2.5.3'),
```
Comment (boegel) [inline easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb]: use latest 2.6?

File: easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    ('fontconfig', '2.11.1'),
+    ('GLib', '2.40.0'),
+    ('libjpeg-turbo', '1.3.1'),
+    ('expat', '2.1.0'),
+    ('cairo', '1.12.18'),
```
Comment (boegel) [inline easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb]: use 1.14.2

File: easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    ('LibTIFF', '4.0.3'),
```
Comment (boegel) [inline easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb]: 4.0.4

File: easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+checksums = ['818c87e31f7562aaa97397d3d0cc20a1']
+
+dependencies = [
+    ('zlib', '1.2.8'),
+    ('libpng', '1.6.17'),  
+    ('freetype', '2.5.3'),
+    ('fontconfig', '2.11.1'),
+    ('GLib', '2.40.0'),
+    ('libjpeg-turbo', '1.3.1'),
```
Comment (boegel) [inline easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb]: 1.4.1

File: easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+sources = [SOURCELOWER_TAR_GZ]
+
+checksums = ['818c87e31f7562aaa97397d3d0cc20a1']
+
+dependencies = [
+    ('zlib', '1.2.8'),
+    ('libpng', '1.6.17'),  
+    ('freetype', '2.5.3'),
+    ('fontconfig', '2.11.1'),
+    ('GLib', '2.40.0'),
```
Comment (boegel) [inline easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb]: 2.45.2?

File: easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+source_urls = ["http://downloads.ghostscript.com/public/"]
+sources = [SOURCELOWER_TAR_GZ]
+
+checksums = ['818c87e31f7562aaa97397d3d0cc20a1']
+
+dependencies = [
+    ('zlib', '1.2.8'),
+    ('libpng', '1.6.17'),  
+    ('freetype', '2.5.3'),
+    ('fontconfig', '2.11.1'),
```
Comment (boegel) [inline easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb]: 2.11.94?

**Comment by boegel** [comment]
@npklein: updating dependency versions in existing easyconfigs is something we try to stay away from, since it may create problems for people who have these installed already...

since there are newer versions for both cairo and freetype, we should use those instead?

**Comment by npklein** [comment]
@boegel Sure we can use the newer versions, but these versions are the ones used in common_deps

**Comment by boegel** [comment]
@npklein cairo and freetype are not in common deps?

**Comment by npklein** [comment]
@boegel ah my bad was confusing something, I'll update the version

**Comment by boegel** [comment]
@npklein: looks good now... how about a test report?

**Comment by npklein** [comment]
@boegel I followed the steps for testing but I get ERROR: Failed to parse configuration options: "Failed to obtain required GitHub token for user 'npklein'", also changed $HOME/.local/share/python_keyring/keyringrc.cfg.

**Comment by boegel** [comment]
@npklein: you should be able to print your GitHub token using:

```
python -c "import keyring; print keyring.get_password('github_token', 'npklein')"
```

Does that work? Make sure you don't share your token with us... :)

**Comment by npklein** [comment]
@boegel it says None but in https://github.com/settings/tokens I have 

EasyBuild easyconfigs PR testing — gist, repo

**Comment by boegel** [comment]
@npklein: that's only part of the story... When you create a token on GitHub, you need to copy-paste it, and then add it in your Python keyring via `keyring.setpassword`, see https://github.com/hpcugent/easybuild/wiki/Review-process-for-contributions#setting-things-up

**Comment by npklein** [comment]
@boegel I did that and I thought it worked if it doesn't give an error, but I guess without giving an error it doesn't mean it worked. It asks for a password and I use my github password there, is that correct?

**Comment by boegel** [comment]
No, the 'password' you need to enter is the token. When it requests a password, just copy-paste the token value and hit enter. Then check via the `get_password` command whether the token is there.

**Comment by npklein** [comment]
@goebel ah I thought 'github_token' had to be substituted by my github token, now it seems to work, thanks!

**Comment by boegel** [comment]
No, `github_token` is the 'key' that EB uses to obtain the token value from your keyring.

**Comment by npklein** [comment]
Test report by @npklein
**FAILED**
Build succeeded for 6 out of 7 (7 easyconfigs in this PR)
Linux SL 6.6, Intel(R) Xeon(R) CPU E5-2697 v2 @ 2.70GHz, Python 2.7.9
See https://gist.github.com/4f6b01fb051b43b5cf7f for a full test report.

**Comment by boegel** [comment]
@npklein: I'm seeing error like this pop up in the log for Ghostscript included in the test report:

```
/apps/software/libjpeg-turbo/1.4.1-goolf-1.7.20/include/jpeglib.h:67:1: error: unknown type name ‘JSAMPLE’
```

I'm not sure what the problem is here...
Maybe this particular version of libjpeg-turbo has bugs that are yet to be resolved.

Can you try with a slightly older version of libjpeg-turbo?
Yes, I know I asked you to use the most recent version, but if it doesn't work... ;-)

**Comment by boegel** [comment]
This looks like a possible explanation: http://sourceforge.net/p/libjpeg-turbo/bugs/76/

**Comment by npklein** [comment]
@boegel I tried with 1.3.1, that also gave an error (don't know if it is
the same). Ill try with an older version later.
Op 23 jun. 2015 20:46 schreef "Kenneth Hoste" <notifications@github.com>:

**Comment by npklein** [comment]
Test report by @npklein
**SUCCESS**
Build succeeded for 7 out of 7 (7 easyconfigs in this PR)
Linux SL 6.6, Intel(R) Xeon(R) CPU E5-2697 v2 @ 2.70GHz, Python 2.7.9
See https://gist.github.com/db9e2c8c1c10a7abdfc6 for a full test report.

**Comment by npklein** [comment]
@boegel I removed the libpng-turbo dependency from the Ghostscript .eb file and it tests successfully. And it's still in the dependency of GLib so it does get installed while installing Ghostscript dependencies (the 1.4.1 version). The .eb files for older versions from Ghostscript do have libpng-turbo in the dependencies, is it very important to keep it in for this version as well?

**Comment by boegel** [comment]
@npklein: libjpeg-turbo is not a dependency of GLib at all? So, by removing it, you're either obtaining a Ghostscript installation with less features, or one that picks up the jpeg library provided by the system (which is not good imho).

I'll take a detailed look myself. It seems like this may just be a missing import in Ghostscript...

**Comment by boegel** [comment]
including `libjpeg-turbo` v1.3.1 works for me, see https://github.com/npklein/easybuild-easyconfigs/pull/1

but, it seems like the library it provides isn't being used... (see comment in PR above)

**Comment by npklein** [comment]
Test report by @npklein
**SUCCESS**
Build succeeded for 7 out of 7 (7 easyconfigs in this PR)
Linux SL 6.6, Intel(R) Xeon(R) CPU E5-2697 v2 @ 2.70GHz, Python 2.7.9
See https://gist.github.com/57b8bd9ac281377e487f for a full test report.

**Comment by npklein** [comment]
@boegel So is the linking of the gs binary isn't linked to the libjpeg.so provided by libjpeg-turob something that we can/should fix or is that a problem in Ghostscript itself and can this eb file be used for installation?

**Comment by boegel** [comment]
Well, it's certainly something we should look into.
This is not a new issue though, existing Ghostscript builds also suffer from it.

Even worse, I'm seeing build that link to two different libjpeg.so libraries:

```
$ ldd $EBROOTGHOSTSCRIPT/bin/gs | grep libjpeg
	libjpeg.so.8 => /user/scratch/gent/vsc400/vsc40023/easybuild_REGTEST/SL6/sandybridge/software/libjpeg-turbo/1.3.1-intel-2014b/lib/libjpeg.so.8 (0x00002ab7fd988000)
	libjpeg.so.62 => /usr/lib64/libjpeg.so.62 (0x00002ab7ff4a4000)
```

File: easybuild/easyconfigs/g/GLib/GLib-2.45.2-goolf-1.7.20.eb [hunk 10]
```diff
@@ -0,0 +1,21 @@
+easyblock = 'ConfigureMake'
+
+name = 'GLib'
+version = '2.45.2'
```
Comment (fgeorgatos) [inline easybuild/easyconfigs/g/GLib/GLib-2.45.2-goolf-1.7.20.eb]: 2.45.4 ? (proposing minor upgrade)

File: easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+checksums = ['829319325bbdb83f5c81379a8f86f38f']
+
+dependencies = [
+    ('zlib', '1.2.8'),
+    ('libpng', '1.6.17'),  
+    ('freetype', '2.6'),
+    ('fontconfig', '2.11.94'),
+    ('GLib', '2.45.2'),
+    ('libjpeg-turbo', '1.3.1'),  # note: libjpeg-turbo 1.4.[01] doesn't work
```
Comment (fgeorgatos) [inline easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb]: there is a mismatch between the above version and the libjpeg provided below

Comment (boegel) [inline easybuild/easyconfigs/g/Ghostscript/Ghostscript-9.16-goolf-1.7.20.eb]: That's OK, all required deps are there, this PR just includes an extra easyconfig for libjpeg-turbo, which is fine.

File: easybuild/easyconfigs/g/GLib/GLib-2.45.2-goolf-1.7.20.eb (hunk 10 above)
Comment (boegel) [inline easybuild/easyconfigs/g/GLib/GLib-2.45.2-goolf-1.7.20.eb]: in fact, this should be 2.44.1; 2.45.x is the 'development' version...

**Comment by boegel** [comment]
@npklein: the linking issue with libjpeg.so can be re-evaluated later; I have a need for Ghostscript 9.16 (with `intel/2015a`, but fine), so let's make some progress and get this merged in

I'll send you a PR to fix the GLib version issue, and then open a companion PR for Ghostscript 9.16 with `intel/2015a`.

**Comment by boegel** [comment]
@npklein: please include https://github.com/npklein/easybuild-easyconfigs/pull/2

**Comment by boegel** [comment]
@npklein: https://github.com/npklein/easybuild-easyconfigs/pull/2

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 13 out of 13 (7 easyconfigs in this PR)
Linux SL 6.6, Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz, Python 2.6.6
See https://gist.github.com/b21f34ad63a8ab080a63 for a full test report.

**Comment by boegel** [comment]
Looking good, going in, thanks @npklein!

**Pull request merged**